Fixed a relatively rare bug and a deprecation warning

omaraltayyan opened the pull request:
Hi.

i was using your library in my project when i came across a small bug while debugging a numpy wrong arrays dimensions exception, so i found a fix for it and thankfully it worked, so i thought i might share it with you to get your thoughts on it.


here's a compressed version of the offending file for the bug


[GLF002591.zip](https://github.com/bjbschmitt/AMFM_decompy/files/2108770/GLF002591.zip)


as for the deprication warning it is caused when using the latest version of the scipy library as of date.

bjbschmitt commented:
Hi Omar, thanks for your feedback!

I will take a look on your commits, test them and update this repository and the pypi one as soon as possible.

omaraltayyan commented:
@bjbschmitt Great!

by the way, have you noticed the many "divide by zero" and "mean of empty slice" warnings during execution? do you think they also happen in the original matlab and are well handled??

bjbschmitt commented:
@omaraltayyan I tested your wave file here, but could not reproduce any of the errors or warnings that you got. could you tell me which python version, scipy version and numpy version are you using? also, are you using any custom parameter value for the pYAAPT.yaapt() function? and also, are using Python in Windows or other OS and are using Anaconda, for example?

by the way, I've tested your file with the original Matlab YAAPT code (version 4.0) and it was not able to finish the pitch extraction. so it seems that at least my python code can handle your problematic file better than the original version.  

(off topic: I've also found a bug in the setup.py file, the package is not installing in python3).

omaraltayyan commented:
@bjbschmitt i am sorry about that

i use python3.6.5 and the latest versions of all the mentioned libraries, as for the sound file, i am doing quite a bit of custom processing like cutting the frequency and slicing the file before passing it to the algorithm, but the only parameter i changed is that i set the fft to 1024 since all my files are at 8000 Hz sampling frequency


anyhow, i think that all you need to reproduce the exception is to manually set `num_voiced_cand` so that the changed code executes (1 or 2), i use pycharm's debugger for this
once you get to this line `spec_pitch[cand_pitch[0, :] > 0] = voiced_pitch[:]` right after the changed code, it will through this exception since the `voiced_pitch` is a 2d array while `spec_pitch` expects a 1D array

bjbschmitt commented:
ok @omaraltayyan , by following your instructions I was able to reproduce the numpy ones array dimension bug. Your solution indeed solves the issue. 

I was also able to reproduce the hanning deprecation warning by running a small script with the extra parameters '-Wd', which makes python to print all warnings. But in this case I'll make some extra modifications regarding the kaiser window. scipy is not throwing yet any warning for it, but I believe that in future versions it will start to recommend using `scipy.signal.windows.kaiser`. So I want to anticipate this change.

The only issue that I still could not reproduce was the "divide by zero" and "mean of empty slice" warnings that you mentioned. Maybe you could give more info about when and where in the code it happens so that I can reproduce it.

For example, the wav sample file that you have sent has sample frequency of 16 KHz and you said that you are cutting the frequency to 8 KHz. Is your script downsampling it?

bjbschmitt commented:
Ok, I'll upload the new version 1.0.8 with the bugs fixed. Regarding the "divide by zero" and "mean of empty slice" warnings, by now I will keep things as they are, since that it seems that the warnings are not affecting the results. Probably I was not able to reproduce them because I'm not using PyCharm. Nevertheless, the AMFM_decompy package is performing much better than the original YAAPT.m file with this "problematic" wav file.

omaraltayyan commented:
@bjbschmitt i am very sorry for the delay, i had a hard test at school and couldn’t reply.

If you wait for a few hours I could get you another file that produces the warnings along with a version of my preprocessing script and the exact numpy and scipy versions.

omaraltayyan commented:
i have removed all preprocessing code to avoid confusion, the only thing that i changed is that i made the fft 1024, all the remaining parameters are the same and the file as whole is passed unchanged to the algorithm

i found two files that throw "mean of empty slice" warnings, i was raising them so the "divide by zero warning" didn't raise so it seems it is raised after the "mean" one

the 1st File: 
[GLF001152.zip](https://github.com/bjbschmitt/AMFM_decompy/files/2127296/GLF001152.zip)

throws the warning from the following:

```
File "C:\Program Files\Python36\lib\site-packages\amfm_decompy\pYAAPT.py", line 408, in yaapt
    final_pitch = dynamic(ref_pitch, ref_merit, pitch, parameters)
  File "C:\Program Files\Python36\lib\site-packages\amfm_decompy\pYAAPT.py", line 710, in dynamic
    mean_pitch = np.mean(best_pitch[best_pitch > 0])
```

the 2nd file:
[GLF001827.zip](https://github.com/bjbschmitt/AMFM_decompy/files/2127300/GLF001827.zip)

throws it at 

```
File "C:\Program Files\Python36\lib\site-packages\amfm_decompy\pYAAPT.py", line 375, in yaapt
    spec_pitch, pitch_std = spec_track(nonlinear_sign, pitch, parameters)
  File "C:\Program Files\Python36\lib\site-packages\amfm_decompy\pYAAPT.py", line 521, in spec_track
    avg_voiced = np.mean(voiced_cand_pitch[0, :])
```

you can reproduce the warnings as exceptions using `np.seterr(all='raise')`

i am using numpy version `1.13.3` and scipy version `1.0.0`


thanks a lot for your efforts.

bjbschmitt commented:
hi @omaraltayyan , sorry for the delay, I am stuck finishing a job for a client. 
thanks for your feedback, I'll take a look on your files, try to reproduce the warnings here and see if I can find a solution for them.

bjbschmitt commented:
hi @omaraltayyan, I've tested your new wav files with the following test script [empty_slice.zip](https://github.com/bjbschmitt/AMFM_decompy/files/2168050/empty_slice.zip), but I'm still not getting the "mean of empty slice" warnings. 

Please check if the script raises the warnings for you. As I said before, maybe is something related with the the operating system or PyCharm (I'm running the script directly in the Linux console, while it seems that you are using PyCharm in Windows).

So if you confirm that this script raises the warnings in your PC, I'll borrow one with Windows here and test it with PyCharm.

omaraltayyan commented:
hi @bjbschmitt your code as it is doesn't reproduce the problem as you have said, but it turns out it's not related to pycharm, since i managed to narrow down the problem, it seems that an error is thrown when using your library with loaded wav files using scipy instead of just passing the file name as you were doing in your script.

here's the code to reproduce the problem, i included a small switch to observe the difference:

`#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import amfm_decompy.pYAAPT as pYAAPT
import amfm_decompy.basic_tools as basic_tools
import numpy as np
import scipy.io.wavfile as wav

# np.seterr(all='raise')

PRODUCE_ERROR = True

def extract_f0_from_file(sound_file_path):
    print("Extracting From: " + sound_file_path)

    if PRODUCE_ERROR:

        (sample_rate, signal) = wav.read(sound_file_path)
        signal_piece_object = basic_tools.SignalObj(signal, sample_rate)
        f0_feature = pYAAPT.yaapt(signal_piece_object)

    else:

        signal_piece_object = basic_tools.SignalObj(sound_file_path)

        f0_feature = pYAAPT.yaapt(signal_piece_object)


extract_f0_from_file('./GLF001152.wav')
extract_f0_from_file('./GLF001827.wav')
`

it seems that only the 2nd file is producing the problem but it's probably not important

Thanks again for your great work, i will try to look for the source of this error if i can.
